### DISCUSSION TITLE
How do I connect an application to a cluster?

### DISCUSSION THREAD

**Discussion by croessner** [OT-CONTAINER-KIT/redis-operator#830] [General]
Hi,

I am new to Kubernetes (still learning on Udemy course) and I installed the redis-operator with the redis-cluster:

```
$ kubectl -n ot-operators get pods                                                                                                                                            09:15:51
NAME                             READY   STATUS    RESTARTS   AGE
redis-cluster-follower-0         1/1     Running   0          2m21s
redis-cluster-follower-1         1/1     Running   0          2m13s
redis-cluster-follower-2         1/1     Running   0          2m4s
redis-cluster-leader-0           1/1     Running   0          3m3s
redis-cluster-leader-1           1/1     Running   0          2m54s
redis-cluster-leader-2           1/1     Running   0          2m45s
redis-operator-c7d844dd4-5lkg8   1/1     Running   0          4m

 $ kubectl -n ot-operators get svc                                                                                                                                             09:16:24
NAME                                TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
redis-cluster-follower              ClusterIP   10.101.17.143    <none>        6379/TCP   2m27s
redis-cluster-follower-additional   ClusterIP   10.100.220.67    <none>        6379/TCP   2m27s
redis-cluster-follower-headless     ClusterIP   None             <none>        6379/TCP   2m27s
redis-cluster-leader                ClusterIP   10.102.52.59     <none>        6379/TCP   3m9s
redis-cluster-leader-additional     ClusterIP   10.108.186.184   <none>        6379/TCP   3m9s
redis-cluster-leader-headless       ClusterIP   None             <none>        6379/TCP   3m9s
webhook-service                     ClusterIP   10.102.217.127   <none>        443/TCP    4m6s
```

So far so good. I already guess that I have to use a headless service, but what do I have to use right now in an application, to connect to this cluster? My expectation would be to define three leaders somehow.

I did not fetch the values for the cluster or the operator. I simply started with the quick start and have no further modifications done.

It would be nice to have some usage information (Also for sentinel/replication which I tried before...)

Thanks in advance

Christian

Update:

I think I must explain, what I am looking for:

I have written a Go application (Open-Source, Nauthilus on Github) that uses go-redis. And go-redis can deal with all kinds of Redis configuration. Therefor I have one code block:

```go
func newRedisClusterClient(redisCfg *config.Redis) *redis.ClusterClient {
	return redis.NewClusterClient(&redis.ClusterOptions{
		Addrs:        redisCfg.Cluster.Addresses,
		Username:     redisCfg.Cluster.Username,
		Password:     redisCfg.Cluster.Password,
		PoolSize:     redisCfg.PoolSize,
		MinIdleConns: redisCfg.IdlePoolSize,
		TLSConfig:    redisTLSOptions(&redisCfg.TLS),
	})
}
```

As you can see, Addrs expects addresses. And now I do not no what to set here. Normally it would be something like:

"addressA:6379 addressB:6379 addressC:6379" and redis-go will do the rest. Do I need only the leader service or also the additional or using the headless and pointing directly to all three pods?

Maybe now my question is more clear :-)

**Comment by drivebyer** [comment]
Hello @croessner! Based on my experience, when using go-redis, we can only specify one service name for `Addrs`, like `redis-cluster-leader`. go-redis will handle the discovery process automatically.
